src/parse_instruction.py

```diff
@@ -352,128 +286,55 @@ def derived(
     ) -> "Instruction":
         return Instruction(mnemonic, args, replace(old.meta, synthetic=True))
 
-    def is_branch_instruction(self) -> bool:
-        return (
-            self.mnemonic
-            in [
-                "j",
-                "b",
-                "beq",
-                "bne",
-                "beqz",
-                "bnez",
-                "bgez",
-                "bgtz",
-                "blez",
-                "bltz",
-                "bc1t",
-                "bc1f",
-            ]
-            or self.is_branch_likely_instruction()
-        )
+    def __str__(self) -> str:
+        args = ", ".join(str(arg) for arg in self.args)
+        return f"{self.mnemonic} {args}"
 
-    def is_branch_likely_instruction(self) -> bool:
-        return self.mnemonic in [
-            "beql",
-            "bnel",
-            "beqzl",
-            "bnezl",
-            "bgezl",
-            "bgtzl",
-            "blezl",
-            "bltzl",
-            "bc1tl",
-            "bc1fl",
-        ]
-
-    def get_branch_target(self) -> JumpTarget:
-        label = self.args[-1]
-        if isinstance(label, AsmGlobalSymbol):
-            return JumpTarget(label.symbol_name)
-        if not isinstance(label, JumpTarget):
-            raise DecompFailure(
-                f'Couldn\'t parse instruction "{self}": invalid branch target'
-            )
-        return label
 
-    def is_jump_instruction(self) -> bool:
-        # (we don't treat jal/jalr as jumps, since control flow will return
-        # after the call)
-        return self.is_branch_instruction() or self.mnemonic == "jr"
+class ArchAsm(abc.ABC):
+    stack_pointer_reg: Register
+    frame_pointer_reg: Register
+    return_address_reg: Register
 
-    def is_delay_slot_instruction(self) -> bool:
-        return self.is_branch_instruction() or self.mnemonic in [
-            "jr",
-            "jal",
-            "jalr",
-        ]
+    base_return_regs: List[Register]
+    all_return_regs: List[Register]
+    argument_regs: List[Register]
+    simple_temp_regs: List[Register]
+    temp_regs: List[Register]
+    saved_regs: List[Register]
 
-    def __str__(self) -> str:
-        args = ", ".join(str(arg) for arg in self.args)
-        return f"{self.mnemonic} {args}"
+    @staticmethod
```

Comment:
no need to make this a staticmethod in the parent interface, should be fine for derived classes to keep state

(unless Python requires it)